eng/build.ps1

```diff
@@ -120,6 +120,9 @@ if ($subset -eq 'help') {
   exit 0
 }
 
+# This tells .NET Core to use the bootstrapped runtime
+$env:DOTNET_ROOT=InitializeDotNetCli -install:$true -createSdkLocationFile:$true
```

Comment:
Why do you need DOTNET_ROOT to be set? We didn't need this before except for Visual Studio and ILLink (?)?

Reply:
On Linux, Crossgen2 executable doesn't launch without DOTNET_ROOT being set (please discussion above in this thread). I guess the clean solution would be to use the dotnet launcher but that's something yet to be implemented in the sdk repo so even after its upgrade to 6.0 it won't come for free.

Reply:
Can we set this environment variable only in the place where crossgen2 is actually invoked? Is that happening in the SDK? Is there a hook to pass environment variable to that invocation directly to avoid global environment variables being set? An MSBuild Exec Task allows to pass in env vars.

Reply:
We don't want to set this environment variable in the root build script as it means that we depend on it throughout the whole build which we actually don't as only crossgen2 needs it. Setting it in the place where crossgen2 is invoked guarantees, that the sfxprojs are buildable when that env var isn't set (i.e. when invoked directly with `dotnet build`, instead of from the root build.cmd/sh).

Reply:
Well, I guess it's mostly a sequencing problem. My initial goal was to just make this work somehow; with help from many people I've managed to achieve that in the current change and we can now discuss the logistics of getting that in. I believe that the problem regarding <code>DOTNET_ROOT</code> is basically a deficiency of the <code>RunReadyToRunCompiler</code> task that should accept a parameter specifying the path to .NET CLI and use the "dotnet" launcher to launch Crossgen2. This logic is absent both in SDK 5.0 and in current main branch of the repo i.o.w. it's not available to us now and won't become immediately available even after we upgrade the SDK repo drop. I can easily sit on this change for a week or two and make the prerequisite changes first in other repos but we need to coordinate the integrations to make sense.

Reply:
In case this is a too oblique answer to your question, the <code>ToolTask</code> base class of <code>RunReadyToRunCompiler</code>, please see

https://github.com/dotnet/sdk/blob/810129ba5b35dbf77f8d6d84171eb7363693551d/src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs#L14

currently doesn't support injecting environment variable overrides into the child process. It is definitely doable, we're doing such a thing in R2RTest and in other places, it's just not there yet. This gets us back to the sequencing thing, it seems to me we're converging towards the conclusion that the best way forward is to fix this cleanly in the SDK repo and simplify this change to just consume it, in such case we'd no longer require global environment overrides like <code>DOTNET_ROOT</code>.

Reply:
Actually the formulation I used is somewhat imprecise - I don't claim ToolTask doesn't support setting environment variables, I have no idea if it does and maybe / likely it does, it's just not used / implemented in the RunReadyToRunCompiler task class derived from ToolTask and so implementing it would require a code change. And my current thinking is that, if we're about to do a code change to the RunReadyToRunCompiler task, it would be easiest to cleanly switch it over to running Crossgen2 using the dotnet launcher and avoid tampering with environment variables.

Reply:
in response to your ToolTask statement, ToolTask already provides a way to specify env vars: https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.utilities.tooltask.environmentvariables?view=msbuild-16-netcore#Microsoft_Build_Utilities_ToolTask_EnvironmentVariables. It's probably just a matter of adding an environment variables property to the RunReadyToRunCompiler task invocation.

Regarding the logistics, I agree that this should be cleanly fixed by the SDK but that doesn't mean that the required environment variable can't be set just in time before the RunReadyToRunCompiler task is invoked.

I was thinking about adding a target that hooks onto `_CreateR2RImages` as a BeforeTargets and sets the environment variable.

Reply:
Good point, fixed.